webots/controllers/RL_Supervisor/rl_supervisor.py

```diff
@@ -61,107 +63,126 @@
 
-        for idx in range(5):
-            if idx == 4:
-                print(f"Sensor[{idx}] = {sensor_data[idx]}")
-            else:
-                print(f"Sensor[{idx}] = {sensor_data[idx]}, ", end="")
-
-        # determine lost line condition
+        # Determine lost line condition
         if all(value == 0 for value in sensor_data):
             self.__no_line_detection_count += 1
         else:
             self.__no_line_detection_count = 0
 
-        # detect start/stop line
-        is_start_stop = all(
-            value >= LINE_SENSOR_ON_TRACK_MIN_VALUE for value in sensor_data
-        )
-
-        # sequence stop criterion debounce no line detection and
-        if (self.__no_line_detection_count >= 20) or is_start_stop:
-            # Stop motors, maximum NO Line Detection Counter reached
-            self.__smp_server.send_data(SPEED_SET_CHANNEL_NAME, struct.pack("2H", 0, 0))
+        # Detect start/stop line
+        if ((sensor_data[SENSOR_ID_MOST_LEFT] >= LINE_SENSOR_ON_TRACK_MIN_VALUE) and
+            (sensor_data[SENSOR_ID_MOST_RIGHT] >= LINE_SENSOR_ON_TRACK_MIN_VALUE)):
+            self.start_stop_line_detected = True
+
+        # Detect Start/Stop Line before Finish Trajectories
+        if (self.start_stop_line_detected is True) and (self.steps < MIN_NUMBER_OF_STEPS):
+            sensor_data = list(sensor_data)
+            sensor_data[SENSOR_ID_MOST_LEFT] = 0
+            sensor_data[SENSOR_ID_MOST_RIGHT] = 0
+            self.start_stop_line_detected = False
+
+        # sequence stop criterion debounce no line detection and start/stop line detected
+        if self.__no_line_detection_count >= 30 or (
+            (self.start_stop_line_detected is True) and (self.steps >= MIN_NUMBER_OF_STEPS)
+        ):
+            self.__agent.done = True
+            self.__no_line_detection_count = 0
+            self.steps = 0
 
-            # SENDING A COMMAND ID SET READY STATUS
-            self.__smp_server.send_data(
-                COMMAND_CHANNEL_NAME, struct.pack("B", CMD_ID_SET_READY_STATE)
-            )
-            self.is_sequence_complete = True
+        # The sequence of states and actions is stored in memory for the training phase.
+        if self.__agent.train_mode:
 
-        else:
-            # [PLACEHOLDER] replace with line sensor data receiver
-
-            # Set left and right motors speed to 1000
-            self.__smp_server.send_data(
-                SPEED_SET_CHANNEL_NAME, struct.pack("2H", 1000, 1000)
-            )
+            # receive a -1 punishment if the robot leaves the line
+            if self.__no_line_detection_count > 0:
+                reward = -1
+            else:
+                reward = self.__agent.determine_reward(sensor_data)
+
+            # Start storage The data after the second received sensor data
+            if self.last_sensor_data is not None:
+                normalized_sensor_data = self.__agent.normalize_sensor_data(self.last_sensor_data)
```

Comment:
Perform preprocessing and transformations at a later stage. This function should be used for data aquisition only. If more input data than just line sensor data is used or the preprocesing operation takes more time than the sampling.